### ISSUE TITLE
Pre-compact upsert formatted input before decoding

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#10819]
### Feature request

Some sources of data are "upsert formatted" in that they have a `key` and `payload` and the intended semantics are that each new message replaces the prior `payload` associated with `key`. In this case, we can pre-thin data we read out that will appear at the same timestamp: multiple messages with the same `key` can be thinned down to just the last `payload`. I think.

This can be done as soon as we have both the timestamp bindings and the message, which can be before e.g. exchanging the messages and decoding the payload, both of which can be non-trivial. Importantly, this only maintains per-timestamp state, which is dropped as soon as the timestamp is complete. The `upsert` operator would still be fully stateful.

There is a follow on discussion about how much work should be performed before or after decoding, but the above optimization seems to be mostly a win in cases where we anticipate some duplication (and not massively harmful in cases we do not).

**Comment by guswynn** [comment]
> multiple messages with the same key can be thinned down to just the last payload.

Yes! this happens here: https://github.com/MaterializeInc/materialize/blob/ebe435da7f949217b359edd7fb953429a18b7bd6/src/dataflow/src/render/upsert.rs#L361-L373 (we don't look at the DD timestamp, we look at the kafka offset, which is more precise)

I think doing just this is a great idea: we can keep upsert stateful, make sure it `evaluates` predicates and so on, but simply feed it significantly less data. I am not 100% on where exactly to place this, and how it works with `DEBEZIUM UPSERT`, so those are things to look into
